pkg/controller/pod_iptables_fip.go

```diff
@@ -0,0 +1,353 @@
+		}
+	} else if c.config.EnableKeepVmIP && isVmPod {
+		if c.isVmPodToDel(p, vmName) {
+			c.delPodAnnotatedIptablesFipQueue.Add(obj)
+			return
+		}
+		if delete, err := appendCheckPodToDel(c, p, vmName, util.VmInstance); delete && err == nil {
+			c.delPodAnnotatedIptablesFipQueue.Add(obj)
+			return
+		}
+	} else {
+		c.delPodAnnotatedIptablesFipQueue.Add(obj)
+		return
+	}
+}
+
+func (c *Controller) runAddPodAnnotatedIptablesFipWorker() {
+	for c.processNextAddPodAnnotatedIptablesFipWorkItem() {
+	}
+}
+
+func (c *Controller) runDelPodAnnotatedIptablesFipWorker() {
+	for c.processNextDeletePodAnnotatedIptablesFipWorkItem() {
+	}
+}
+
+func (c *Controller) processNextAddPodAnnotatedIptablesFipWorkItem() bool {
+	obj, shutdown := c.addPodAnnotatedIptablesFipQueue.Get()
+	if shutdown {
+		return false
+	}
+	err := func(obj interface{}) error {
+		defer c.addPodAnnotatedIptablesFipQueue.Done(obj)
+		var key string
+		var ok bool
+		if key, ok = obj.(string); !ok {
+			c.addPodAnnotatedIptablesFipQueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+		if err := c.handleAddPodAnnotatedIptablesFip(key); err != nil {
+			c.addPodAnnotatedIptablesFipQueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+		}
+		c.addPodAnnotatedIptablesFipQueue.Forget(obj)
+		return nil
+	}(obj)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+	return true
+}
+
+func (c *Controller) processNextDeletePodAnnotatedIptablesFipWorkItem() bool {
+	obj, shutdown := c.delPodAnnotatedIptablesFipQueue.Get()
+	if shutdown {
+		return false
+	}
+	err := func(obj interface{}) error {
+		defer c.delPodAnnotatedIptablesFipQueue.Done(obj)
+		var pod *v1.Pod
+		var ok bool
+		if pod, ok = obj.(*v1.Pod); !ok {
+			c.delPodAnnotatedIptablesFipQueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected pod in workqueue but got %#v", obj))
+			return nil
+		}
+		if err := c.handleDeletePodAnnotatedIptablesFip(pod); err != nil {
+			c.delPodAnnotatedIptablesFipQueue.AddRateLimited(obj)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", pod.Name, err.Error())
+		}
+		c.delPodAnnotatedIptablesFipQueue.Forget(obj)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+	return true
+}
+
+func (c *Controller) handleAddPodAnnotatedIptablesFip(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+	cachedPod, err := c.podsLister.Pods(namespace).Get(name)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	if cachedPod.Annotations[util.FipEnableAnnotation] != "true" {
+		// not enable fip
+		return nil
+	}
+	if cachedPod.Annotations[util.FipNameAnnotation] != "" {
+		// fip aleady ok
+		return nil
+	}
+	pod := cachedPod.DeepCopy()
```

Comment:
for read only operations no need for deepcopy

Reply:
fixed